### PR TITLE
Fix exam quiz submissions not being saved properly

### LINKED ISSUE #1770: Exam mode: Progress is sometimes not stored for multiple choice questions
### Describe the bug
Students reported that progress was lost on some exercises when they reloaded the page, even when they clicked on the "submit & continue" button and when they worked on the exercise more than 60 seconds ago before they reloaded the page.

#### To Reproduce
1. participate in exam
2. work on multiple choice questions
3. Click on "submit & continue" and additionally wait > 60 seconds
4. Some of the progress (e.g. check marks in MC questions) may be lost

#### Expected behavior
Progress should not be lost. 
Additionally, it would be nice that, when some of the current state is not stored, it were indicated to the user.

#### Additional context
Some users reported that at the end of the exam, their summary did not contain some of the work they did on the exercises.
I am not sure whether this is related and e.g. caused by progress that has not been stored.

### PR DESCRIPTION
**Pull request by MaisiKoleni**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This resolves #1770 

### Description
- force save submission when the user clicks on save
- implement can deactivate
- add TODO for generic message
- small code refactoring

### Steps for Testing
Prerequisite: An Exam was properly set up with multiple choice questions and is running.
- Check that in the exam view, a warning is displayed when you try to reload the page.
- If the network connection is lost and you try to save ...
    * by waiting 60 seconds after a change
    * by manually clicking on "Safe and Continue" _(this will force a save, regardless of)_
    * by switching the exercise in the navigation bar
    
    a warning shows up, that there submission could not be saved.
- if there is no sign of it being not saved, it is present after reload *or* the end of the exam (summary)

### CODE REVIEW

File: src/main/webapp/app/exam/participate/exam-participation.component.ts [hunk 1]
```diff
@@ -233,13 +255,15 @@ export class ExamParticipationComponent implements OnInit, OnDestroy {
...
+     * @param force is set to true, when the current exercise should be saved (even if there are no changes)
      */
-    triggerSave(intervalSave: boolean) {
+    triggerSave(intervalSave: boolean, force: boolean) {
         // before the request, we would mark the submission as isSynced = true
         // right after the response - in case it was successful - we mark the submission as isSynced = false
         this.autoSaveTimer = 0;
 
-        if (this.currentSubmissionComponent?.hasUnsavedChanges()) {
+        if (force || this.currentSubmissionComponent?.hasUnsavedChanges()) {
```
Comment (TobiasPr) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:265]: Suggestion: this.currentSubmissionComponent && force ||

File: src/main/webapp/app/exam/participate/exam-participation.component.ts [hunk 2]
```diff
@@ -256,31 +280,32 @@ export class ExamParticipationComponent implements OnInit, OnDestroy {
         });
 
         // if no connection available -> don't try to sync
-        if (!this.disconnected) {
+        if (force || !this.disconnected) {
```
Comment (TobiasPr) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:283]: why do we try to update? is the this.disconnected that unreliable?

File: src/main/webapp/app/exam/participate/exam-participation.component.ts (hunk 1 above)
Comment (krusche) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts:265]: Why would you change the order?

File: src/main/webapp/app/exam/participate/exam-participation.component.ts (hunk 2 above)
Comment (krusche) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts:283]: Yes, it’s unreliable and if the student clicks on save we should at least try it.
Also the connection status shows the Websocket status, a Rest call could still work even if the Websocket is not connected (for whatever reason)

File: src/main/webapp/app/exam/participate/exam-participation.component.ts (hunk 1 above)
Comment (krusche) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts:265]: How can it happen that the currentSubmissionComponent is not available?

**Comment by jpbernius** [comment]
I tested this on TS3 and this looks very good.

- When reloading the page, I get warned and can cancel and save before reloading.
- clicking save / continue & save will trigger a network call
- timer based saved also works (after which I can reload the page and my answers are still there.

The only thing I noticed is when I switch between exercises using the navigation elements on top, it does not save, reloading leads to data loss (if I ignore the warning).
As I see this, this is on purpose?
```ts
changeExercise(i, false)
```

**Comment by TobiasPr** [comment]
> I tested this on TS3 and this looks very good.
> 
> * When reloading the page, I get warned and can cancel and save before reloading.
> * clicking save / continue & save will trigger a network call
> * timer based saved also works (after which I can reload the page and my answers are still there.
> 
> The only thing I noticed is when I switch between exercises using the navigation elements on top, it does not save, reloading leads to data loss (if I ignore the warning).
> As I see this, this is on purpose?
> 
> ```ts
> changeExercise(i, false)
> ```

That is really weird, because they are called in the same way 
periodic save -> `this.triggerSave(true, false);`
exercise change -> `this.triggerSave(false, false);` (but the first flag is ignored as far as I can see)

**Comment by krusche** [comment]
@jpbernius if you have changes, this should also save. 
The guard does however not check if you actually have changes. It always warns you

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.html [hunk 3]
```diff
@@ -47,6 +47,7 @@ <h2>
                 *ngIf="question.type === MULTIPLE_CHOICE"
                 [question]="question"
                 [(selectedAnswerOptions)]="selectedAnswerOptions[question.id]"
+                [fnOnSelection]="onSelectionChanged.bind(this)"
```
Comment (jpbernius) [inline src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.html:50]: Why do we need `.bind(this)` here? Is `onSelectionChanged()` not sufficient?

**Comment by jpbernius** [comment]
Do we also need change detection for the Online Code Editor?

File: src/main/webapp/i18n/de/exam.json [hunk 4]
```diff
@@ -49,7 +49,9 @@
             "submitOtherExercise": "Speichern & Weiter",
             "submitLastExercise": "Speichern",
             "greeting": "Hallo {{ fullName }}, willkommen zu \"{{ title }}\"",
-            "points": "Punkte"
+            "points": "Punkte",
+            "saveSubmissionError": "Die Änderungen konnten nicht gespeichert werden! Bitte stelle sicher, dass du online bist und speichere nochmal.",
+            "pendingChanges": "Alle ungespeicherten Änderungen gehen beim Verlassen der Klausur verloren. Möchtest du sicher fortfahren?"
```
[the remark this answers is not in the document]
Comment (TobiasPr) [reply src/main/webapp/i18n/de/exam.json:54]: Done

**Comment by schultek** [comment]
Seems to work well. Could not get any unsaved or saved changes to disappear, except when reloading the page and dismissing the warning.
However it might be a bit confusing that the warning shows regardless if there are any changes or not.
The footer still needs to be adjusted to say 30s from what i've seen in the files.

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.html (hunk 3 above)
Comment (krusche) [reply src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.html:50]: just copied from the other part. As it is working, I don't want to touch this ;-)

**Pull request merged**